### ISSUE TITLE
engine improvement: "confusing" moves in a losing position

### ISSUE THREAD

**Issue by Twixtfanatic** [stevens68/twixtbot-ui#68]
Here's a request that would not necessarily involve generating a new engine from scratch.

Based on BonyJordan's explanation of the bot's behavior, if it finds that a specific move loses, it will try the next move on a list, then the next, and so on without regard to the time limit it is working under. So for example it might reach the 50th best move, and be part way through the analysis when it runs out of time. It hasn't realized yet that choice 50 also loses, so it makes that move.

I propose that instead of doing that, for each move that loses, spend some time evaluating how difficult it would be for the opponent to find the proper winning continuation against that move. Ideally, there should be only one way to win against this move. Of course it would take too much time to guarantee there is only one way to win against it, but some sort of scoring system would be a lot better way to generate confusing moves, thereby providing a stronger opponent against humans. Thank you for your attention.

**Comment by stevens68** [comment]
I think I'd need one or more example games when this is happening - including the number of trials used (I heard at littlegolem it's 50.000) and then I could reproduce it with twixtbot-ui and dig deeper.

**Comment by Twixtfanatic** [comment]
One such game is at [(http://twixt-commentator.duckdns.org/game/2170049)] The crucial position is after 26.G10. white 27.G11 would have put up a much stronger defense than D10. As far as I can tell, The only way for black to win would have been 28.H4. This was using the bot's standard time control, about 10 minutes per move. Maybe BonyJordan would know how many trials that is.

I will look among the commented bot vs bot games for more examples.

**Comment by Twixtfanatic** [comment]
Copying and pasting from Jordan's own comment in my previous link, here are three snapshots of what the machine was thinking during the game.

```
0.678 |23.l19 24.i9 25.h10 26.k16 27.n14 28.j14 29.k11 30.n19 31.q17 32.g10 33.j9
0.200 |25.f9 26.j4 27.k17 28.e4 29.i5 30.i6 31.d5 32.c5 33.f4 34.g3 35.j3 36.i2 37.h3 38.k3 39.l4
-0.590 |27.d10 28.k16 29.j16 30.n19 31.f20 32.c18 33.r13 34.m21 35.j20
```
When it made move 23, it thought it was in a very favorable position.
Then after 24.H8 its evaluation decreased significantly, but it made a good move 25.F9.
Then after 26.G10 it made the terrible move D10.

Maybe your settings fixed whatever the problem was, but I have another favor to ask. Could you please set up the position after 27.G11 28.H4 and tell me how your machine evaluates that? Apparently it still thinks it has a favorable position after G11 yes? I'd like to know the point where it realizes it is losing, and how it responds then. For example one crucial line is
29.i10 30.e3 31.g6 32.f5 33.k6 34.j5 35.m5 36.l10 37.n14 38.k7 but of course your bot may vary. Thank you.

**Comment by stevens68** [comment]
I ran the bot with 10.000 trials:

  |   | eval | bot would play |   | eval after that move
-- | -- | -- | -- | -- | --
27 | G11 | 0,890 |   |   |  
28 | H4 | 0,738 | 29 | J5 | 0,838
29 | I10 | 0,431 | 30 | F3 | -0,116
30 | E3 | 0,738 | 31 | C4 | 0,736
31 | G6 | 0,831 | 32 | K16 | 0,308
32 | F5 | 0,601 | 33 | C4 | 0,600
33 | K6 | 0,306 | 34 | L4 | -0,632
34 | J5 | 0,546 | 35 | L6 | -0,400
35 | M5 | -0,261 | 36 | K7 | -0,587
36 | L10 | 0,854 | 37 | C5 | 0,914
37 | N14 | -0,286 | 38 | K7 | -0,872
38 | K7 | -0,872 |   |   |  

column eval has the immediate evaluation (before MCTS) which is positive until 34.J5 - but what counts is twixtbot's decision after 10.000 trials (I was too impatient to set it to 50.000 trials).
At move 33 and 34, after 10.000 trials, it recognizes that the situation is actually negative.  
Evaluation after 36.L10 is something I will have a closer look. Even after 10.000 trials, it thinks it has a very favorable position.
interestingly, after each move twixtbot would move differently than your line suggests - except 38.K7

**Comment by stevens68** [comment]
Bot's line is ....37.C5, 38.K7, 39.J8, 40.C2, 41.L7, 42.K16, 43.J16, 44.N19
After 44.N19 it has a negative evaluation

**Comment by BonyJordan** [comment]
The bot uses 50,000 trials on little golem (slow uses 2 million)

**Comment by BonyJordan** [comment]
If you want to understand how things are going over the course of those 50,000 trials, what I would do is go into `nnmcts.visit_node` look for where `U` is for the root (i.e. when `top and self.smart_root` is `True`.  Print out what `Q,P,N,U` are for each move.

**Comment by stevens68** [comment]
Thanks for jumping in Jordan.

to find out why twixtbot@littlegolem chose D10 but twixtbot on my desktop chooses G11 we'd have to compare 
all the params that could lead to a different Q, P, N, U values during MCTS. I meanwhile suspect these 2 settings might be the reason:
- random rotation: "off" in my settings, defaults to "on" in twixtbot repo
- smart_root: "off" in my settings; defaults to "on" in twixtbot repo    (to be honest, I didn't really understand what it does  ;-)

other params that might be different are:
- add_noise: 0.0 in my settings 
- temperature: 0.0 in my settings 
- cpuct: 1.0 in my settings
- network: is it the same network used at littlegolem as the one shared in twixtbot repo?

**Comment by BonyJordan** [comment]
`random_rotation` means you randomly choose whether to mirror on the X axis or Y axis (or both or neither) before going to the net.  If the net were perfect it wouldn't make any difference, but as it is, you'll get different evaluations.  If you want the bot to not play the exact same game each time, then `True` is a good choice.  If you want to get the exact same evaluation every time you evaluate the exact same position, then `False` is a good choice.

`smart_root = False` is ... not recommended.  The move TB chooses is the move that got the most visits.  If after 10,000 nodes you visited move X 5,300 times and Y 4,700 times, then sure, you choose X.  But that means you completely wasted 599 visits.  Once you've visited X 4,701 times, an additional visit to X has no chance to change its mind, so why bother?  If you visit Y a second time and get to 4,701 and 4,701 now you are alllowed to visit X again; and as long as X keeps looking better than Y, you will.

**Comment by stevens68** [comment]
... I noticed yesterday that the 4 evaluations can differ quite significantly. I think I indeed switched it off to get the exact same value for the same position. Now I'm thinking about a 3rd option "average"  that averages over all 4 values (although it will slow down the bot).

Regarding smart_root: Now I understand why I saw so many close "visit-races" when I initially played around with twixtbot/MCTS. I get the point of smart_root and I think we should make it optional in twixtbot-ui, so everybody can play around with it. From a gut-feeling I'd probably keep it off and trust in the power of the U value, i.e. whichever move has the best U, deserves to be visited - including the leading move. 

Thanks for clarifying!

**Comment by stevens68** [comment]
smart_root implemented in PR#82. Couldn't reproduce 27.D10 though  in http://twixt-commentator.duckdns.org/game/2170049.
(smart_root=true, random_rotation=true, trials=50000). But will try again. Maybe random_rotation will lead to different results...

**Comment by stevens68** [comment]
closing the issue.

Thanks for the discussion that led to enabling "smart_root" and rotation options in twixtbot-ui!

Regarding the initial suggestion by Twixter: 
I didn't see any explicit code in twixtbot handling a losing position. In any situation in takes the best move it finds - which should be the one that makes it as hard a possible for the opponent to win.

Unfortunately wasn't able to find why twixtbot didn't play G11 in  [(http://twixt-commentator.duckdns.org/game/2170049)]
- which it did when I played it locally with twixtbot-ui. (Maybe it just had a bad day ;-)
